File: smart_freight.client/src/types/trip.ts
export type TripLifecycleStatus = 'Waiting' | 'Active' | 'Cancelled' | 'Done';

export type TripSummary = {
    id: string;
    truckId: string;
    truckName: string;
    driverId: string;
    driverName: string;
    status: TripLifecycleStatus;
    totalPlannedDistance: number;
    createdAt: string;
    doneAt?: string | null;
};

export type TripCargoItemRequest = {
    productId: string;
    quantity: number;
};

export type TripPlanRequest = {
    truckId: string;
    driverId: string;
    startNodeId: string;
    cargoItems: TripCargoItemRequest[];
    stopLocationIds: string[];
    status?: Exclude<TripLifecycleStatus, 'Done'>;
};

export type TripStop = {
    id: string;
    stopLocationId: string;
    stopName: string;
    stopOrder: number;
    status: string;
};

export type TripRouteStep = {
    stepOrder: number;
    fromNodeId: string;
    toNodeId: string;
    edgeWeight: number;
    cumulativeWeight: number;
};

export type TripPlanResponse = {
    tripId: string;
    totalWeightKg: number;
    totalPlannedDistance: number;
    stops: TripStop[];
    routeSteps: TripRouteStep[];
};

export type TripCargo = {
    productId: string;
    productName: string;
    quantity: number;
    totalWeightKg: number;
};

export type TripDetails = {
    id: string;
    truckId: string;
    truckName: string;
    driverId: string;
    driverName: string;
    status: TripLifecycleStatus;
    totalPlannedDistance: number;
    createdAt: string;
    doneAt?: string | null;
    stops: TripStop[];
    cargoItems: TripCargo[];
};

export type TripStatusUpdateRequest = {
    status: Exclude<TripLifecycleStatus, 'Done'>;
};

export type TripStopStatusUpdateRequest = {
    status: string;
    notes?: string | null;
};

File: Smart_Freight.Server/Models/Trip.cs
namespace Smart_Freight.Server.Models;

public class Trip
{
    public Guid Id { get; set; }
    public Guid TruckId { get; set; }
    public Guid DriverId { get; set; }
    public TripStatus Status { get; set; } = TripStatus.Waiting;
    public decimal TotalPlannedDistance { get; set; }
    public decimal? TotalPlannedCostOrTime { get; set; }
    public string CreatedByUserId { get; set; } = string.Empty;
    public DateTimeOffset CreatedAt { get; set; } = DateTimeOffset.UtcNow;
    public DateTimeOffset? DoneAt { get; set; }

    public Truck Truck { get; set; } = null!;
    public Driver Driver { get; set; } = null!;
    public ApplicationUser CreatedByUser { get; set; } = null!;
    public ICollection<TripStop> Stops { get; set; } = new List<TripStop>();
    public ICollection<TripCargoItem> CargoItems { get; set; } = new List<TripCargoItem>();
    public ICollection<TripRouteStep> RouteSteps { get; set; } = new List<TripRouteStep>();
}
